server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/service/EventDeadLettersService.java

```diff
@@ -19,50 +19,115 @@
 import org.apache.james.mailbox.events.Event;
+import org.apache.james.mailbox.events.EventBus;
 import org.apache.james.mailbox.events.EventDeadLetters;
 import org.apache.james.mailbox.events.Group;
+import org.apache.james.task.Task;
+import org.apache.james.webadmin.dto.ActionEvents;
 
 import com.github.steveash.guavate.Guavate;
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
 
 public class EventDeadLettersService {
     private final EventDeadLetters deadLetters;
+    private final EventBus eventBus;
     private final EventSerializer eventSerializer;
 
     @Inject
-    public EventDeadLettersService(EventDeadLetters deadLetters, EventSerializer eventSerializer) {
+    public EventDeadLettersService(EventDeadLetters deadLetters, EventBus eventBus, EventSerializer eventSerializer) {
         this.deadLetters = deadLetters;
+        this.eventBus = eventBus;
         this.eventSerializer = eventSerializer;
     }
 
-    public ImmutableList<String> listGroups() {
-        return deadLetters.groupsWithFailedEvents()
+    public List<String> listGroupsAsString() {
+        return listGroups()
             .map(Group::asString)
             .collect(Guavate.toImmutableList())
             .block();
     }
 
-    public ImmutableList<String> listGroupEvents(Group group) {
-        return deadLetters.failedEventIds(group)
+    private Flux<Group> listGroups() {
+        return deadLetters.groupsWithFailedEvents();
+    }
+
+    public List<String> listGroupsEventIdsAsStrings(Group group) {
+        return listGroupEventIds(group)
             .map(Event.EventId::getId)
             .map(UUID::toString)
             .collect(Guavate.toImmutableList())
             .block();
     }
 
+    private Flux<Event.EventId> listGroupEventIds(Group group) {
+        return deadLetters.failedEventIds(group);
+    }
+
+    private Mono<List<Event>> listGroupEvents(Group group) {
+        return listGroupEventIds(group)
+            .flatMap(eventId -> getEvent(group, eventId))
+            .collect(Guavate.toImmutableList());
+    }
+
     public String getSerializedEvent(Group group, Event.EventId eventId) {
-        return deadLetters.failedEvent(group, eventId)
+        return getEvent(group, eventId)
             .map(eventSerializer::toJson)
             .block();
     }
 
+    private Mono<Event> getEvent(Group group, Event.EventId eventId) {
+        return deadLetters.failedEvent(group, eventId);
+    }
+
     public void deleteEvent(Group group, Event.EventId eventId) {
         deadLetters.remove(group, eventId).block();
     }
+
+    public Task createActionOnEventsTask(ActionEvents action) {
+        Map<Group, List<Event>> groupsWithEvents = new HashMap<>();
+
+        listGroups()
+            .map(group -> listGroupEvents(group).map(events -> groupsWithEvents.put(group, events)))
```

Comment:
map should not perform side effects on a local variable, you need a collect/reduce instead.